general/globals.hpp

```diff
@@ -125,5 +125,17 @@ void SetGlobalMPI_Comm(MPI_Comm comm);
 #define MFEM_DEPRECATED
 #endif
 
+#if !defined(NDEBUG)
+#include <cassert>
+#define MFEM_ABORT_KERNEL assert(0);
+#endif
+
+#if defined(NDEBUG) && defined(__CUDA_ARCH__)
+#define MFEM_ABORT_KERNEL asm("trap;");
+#endif
+
+#if defined(NDEBUG) && !defined(__CUDA_ARCH__)
+#define MFEM_ABORT_KERNEL abort();
+#endif
```

Comment:
I few things:
* Do not use `NDEBUG` (which means NO-debug), instead use `MFEM_DEBUG` (which means YES-debug).
* `MFEM_ABORT_KERNEL` should be made consistent with `MFEM_ABORT` and abort always, independent of debug macros like `NDEBUG`/`MFEM_DEBUG`.
* It is best to move these definitions with other similar macros like `MFEM_ABORT` in `error.hpp`.

Reply:
Should I just do `#define MFEM_ABORT_KERNEL abort();` without checking for  `__CUDA_ARCH__`? Will that work in every case?

Reply:
I don't think you can call `abort()` on GPU. That's why when compiling the device code we want to use `asm("trap;");`. When compiling the file with `nvcc` for device, `__CUDA_ARCH__` is defined. In all other cases, `__CUDA_ARCH__` is not defined.

In the latter case, you can replace `abort();` with `MFEM_ABORT("kernel abort");`. Even better may be to add an argument to `MFEM_ABORT_KERNEL` that is passed down to `MFEM_ABORT` when `__CUDA_ARCH__` is not defined and it is ignored in the other case.

Reply:
``abort()``  works on AMD GPUs